### ISSUE TITLE
Problem simulation ~250K Atoms

### ISSUE THREAD

**Issue by afarnudi** [openmm/openmm#3056]
Short version:
When scaling up a simulation to 246_856 Atoms (works perfectly for 50K Atoms) I get:
OpenCL internal error: CL_MEM_OBJECT_ALLOCATION_FAILURE error executing CL_COMMAND_WRITE_BUFFER on Tesla P100-PCIE-16GB (Device 0).
EXCEPTION: Error uploading array interactionGroupData: clEnqueueWriteBuffer (-4)
I get the same error with:
RTX 2080 SUPER (8GB), P100-PCIE-12GB, GTX 1080 (8GB) 

Long version:
I have ~250K atoms representing 8 coarse grained chromatin chains (I will need to add extra 50K as well). The chains are connected by CustomBondForces (Kremer-Grest model). They have CustomAngleForces, and all particles (except the ones connected with bonds) have NonBondedCustomForce (the WCA interaction).

The system works excellently when I setup all the atoms but turn off the NonBondedCustomForces. But the issue pasted above arises when I switch on the non bonded interactions. I should add that everything works for a smaller system (around 50K particles).

I first tried it on my 2020 Macbook Pro. It told me that my GPU needed around 2GB of RAM (I have 1.5 GB) so I used  big GPUs (listed above). But it failed (a copy of the errors are in the short version). I should note that it works on the CPU platform. But I want to use GPU.

I noticed that a similar issue was raised #2696 in May 2020. But I did not understand if it was solved and available in the latest release 7.5.0.

Cheers,
Ali

**Comment by peastman** [comment]
Generally it should have no problem simulating systems that large, but of course it depends on the details of your custom forces.  Can you provide a reproducible test case?  If you just provide a serialized System, plus a PDB file with the initial coordinates, that should be fine.

```python
with open('system.xml', 'w') as output:
  output.write(XmlSerializer.serialize(system))
```

**Comment by afarnudi** [comment]
Hi Peter

Thank you for the quick reply.
Here is a PDB of the initial coordinates (I plan to do a minimisation before starting the simulation).
[initial_coordinates.pdb.zip](https://github.com/openmm/openmm/files/6118783/initial_coordinates.pdb.zip)

I realise that the PDB alone is not useful. I use the C++ API. Is there a method similar to "XmlSerializer" in there?

**Comment by peastman** [comment]
The corresponding C++ code should be something like this.

```c++
ofstream output("system.xml");
XmlSerializer::serialize<System>(&system, "System", output);
output.close();
```

**Comment by afarnudi** [comment]
Ok, I hope I did everything right. 
This is the system right before I use it to create a context + a PDB of the initial positions.

[System_PDB.zip](https://github.com/openmm/openmm/files/6118942/System_PDB.zip)

**Comment by peastman** [comment]
You have 36 CustomNonbondedForces, all of which are nearly identical???  I'm not sure what the goal of that is, but I'm pretty sure there's a more efficient way of doing it!  Can you describe what you're trying to compute?

**Comment by afarnudi** [comment]
The system is suppose to represent 8 chromatin chains that will later be wrapped by a coarse grained membrane (2D triangular mesh). 

There are a few reasons why I went with multiple custom forces:

(1) I wanted to have the freedom to assign "force groups" to some of them so I could monitor the evolution of the individual "Forces". For instance when I want to measure the force applied by any of the chains on the membrane (not present in the system I uploaded).

(1.1) I am aware of the 31 limit imposed by the kernel when assigning force groups but I deduced that as long as I keep all forces in one group and assign force groups to a selected few, I can work around the limitation.

(2) I weighted defining one custom force and using PerParticleParameter declaration against  multiple forces with GlobalParameters. I went with the other since it sounded more memory efficient and it was inline with the point I made in (1)

I am PhD student in Physics. I am writing the code (in C++) I use for my thesis in the form of an open source software package with OpenMM as the main engine so that future students in our group (Softmatter group at Sharif University of Technology, Tehran, Iran) or anyone else, can use it to tackle a wide range of problems.

**Comment by peastman** [comment]
Using a single force with a per-particle parameter will be *far* more efficient.  Not just in memory, but speed as well.

**Comment by afarnudi** [comment]
Thank you for the info. I will setup a new system.

Am I to conclude that the per-particle parameter approach is always the better choice? Even if I have multiple forces?

**Comment by peastman** [comment]
It's hard to give a universal rule.  There's an overhead to each additional force.  There also is an overhead to each per-particle parameter.  In this case, you're choosing between adding one parameter or 35 forces, so that's a pretty extreme case.  In other situations it might be less clear, so you would need to try both to find out which is faster.

Using interaction groups also makes the calculation less efficient.  If you only need to compute a small fraction of interactions, it can still end up being faster.  But if you have a bunch of interaction groups that collectively add up to computing all interactions in the whole system, that will be a lot slower than not using interaction groups.

**Comment by afarnudi** [comment]
Dear Peter

The issue was resolved thanks to your suggestions.

I am trying to work around a minimisation issue I now face. The issue #4554 helped a lot in understanding how the minimisation works. If you prefer, we can hold this one open and I will update you if the new issue is resolved?

**Comment by peastman** [comment]
There is no issue 4554.  Did you mistype the number?  Anyway, let's close this.  You can open a new issue if you run into other problems.

**Comment by afarnudi** [comment]
I think I need more help. The system I made using your advice works, but as soon as I slightly stray from it I get the following error:

OpenCL internal error: CL_MEM_OBJECT_ALLOCATION_FAILURE error executing CL_COMMAND_WRITE_BUFFER on GeForce RTX 2080 SUPER
EXCEPTION: Error uploading array interactionGroupData: clEnqueueWriteBuffer (-4)

What I have in my system:
271,858 Atoms
[initialCoordinates.xyz.zip](https://github.com/openmm/openmm/files/6213109/initialCoordinates.xyz.zip)

When I define a single CustomNonBondedForce between all of them, I do not seem to run into trouble. Except that it is super slow (step size 0.01 for 1000 steps it tool 13 hours) and I think it is because I did not setCutoffDistance. Here is the system:
[system_OK.xml.zip](https://github.com/openmm/openmm/files/6213115/system_OK.xml.zip)

I want to split them into 2 groups to set separate CutoffDistances (because I assume that by setting the CutoffDistances  OpenMM will  make neighbour lists and speed everything up for me):
group 1      246,856 Atoms
group 2        25,002 Atoms

I want:
group 1 to see group 2 with  Cuttoff_1,
and  group 1 see itself with Cuttoff_2

So I declared 2 forces (very small difference) and tried using the addInteractionGroup feature. But it fails with the above error. Here is the failing system:
[system_fail.xml.zip](https://github.com/openmm/openmm/files/6213129/system_fail.xml.zip)

I am afraid that I haven't got the hang of using OpenMM efficiently.
Thank you in advance.

**Comment by peastman** [comment]
I really wouldn't use interaction groups.  As I noted above, they make the calculation less efficient (both in time and memory).  If you only need a small fraction of the interactions in the system, they can still be beneficial.  But if the groups add up to computing all interactions, you're much better off not using them.

Not using a cutoff will definitely cause it to be very slow.  With 271,858 atoms, a single force evaluation without a cutoff means computing almost 37 billion pairwise interactions!  Enable a reasonable cutoff and it should get a lot faster.
